File: LongestPallindrome_LC409.java
/**
 * Maintain a HashSet to keep track of characters and count to keep track of length of pallindrome
 * When a new char is seen, add it to set
 * When it already exists in set, you found a pair so remove from set and do count + 2
 * After wh
 */

// Time Complexity: O (N) N: length of string
// Space Complexity: O(N) worst case all characters are unique
// Did this code successfully run on Leetcode : Yes
import java.util.HashSet;
import java.util.Set;

public class LongestPallindrome_LC409 {
    public int longestPalindrome(String s) {
        // Handle empty input
        if (s == null || s.length() == 0)
            return 0;

        int count = 0;                              // Keep track of length of pallindrome
        Set<Character> charSet = new HashSet<>();
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (!charSet.contains(c))               // If c is not in set, add it
                charSet.add(c);
            else {                                  // If c is in set, remove it and increment count by 2 so as to form pallindrome cc
                charSet.remove(c);
                count += 2;
            }
        }
        return charSet.size() > 0 ? count + 1 : count;      // If there is an extra char in hashset, it can be center element
    }
}

File: longestPallindrome-LC 409.py
# Time Complexity = O(n)
# Space Complexity = O(1)


class Solution:
    def longestPalindrome(self, s: str) -> int:
        if len(s) == 0 or s == None:
            return 0
    
        palSet = set()      # For avoiding duplicate values, using set()
        length = 0
    
        for i in range(len(s)):
            char = s[i]

            # If char is is not in set, add it
            if char not in palSet:
                palSet.add(char)
            # If char is already present, increment the length by 2, and remove the already present char from set
            else:
                length += 2
                palSet.remove(char)
        
        # If set is empty, or len(set) == 0, it means that even number of characters were matched perfectly. Return length
        if not palSet:
            return length
        # If set isn't empty, that means atleast one unmatched character would be there. so max length becomes length + 1
        else:
            return length + 1
    
        
